(not ready) publish/ - re-translations (#1738)
cocos-creator/creator-docs: en/editor/publish/native-options.md

# Publish to Native Platforms

Click **Project -> Build** in the menu bar to open the Build panel.

The native platforms that can be selected currently include **Android**, **iOS**, **Mac**, and **Windows**. Among them, the options for publishing to **iOS**, **Mac**, and **Windows** only appear when using the corresponding operating system.

![native platform](publish-native/native-platform.png)

> ## Environment configuration

Reviewer: Please correct all the titles if you've decided the standard.

> **Notes**:
>
> **About the source code engine**
>
> The Cocos2d-x engine includes source code engines, and the scope of their application is:
> 1. When the source code engine first builds and compiles a project, it takes a long time to compile C++ code. Depending on the computer configuration, this time may be 5-20 minutes. For the same project, after it has been compiled once, the time required for the next compilation will be greatly reduced.
> 2. The project built by the source code engine is compiled and run using the native development environment (such as Android Studio, Xcode and other IDEs), which can be debugged and error captured.

The `resources\3d\cocos2d-x-lite` folder under the installation directory of Cocos Creator already contains the built-in Cocos2d-x source code engine. If you need to customize the engine, please refer to the [Engine Customization Workflow](../../advanced-topics/engine-customization.md) documentation for details.

#### Resource Server Address

> When the package body is too large, the resource can be uploaded to the resource server and downloaded through the network. This item is used to fill in the address where the resource is stored on the remote server. Developers need to manually upload the `remote` folder in the release package directory to the filled resource server address after construction. For details, please refer to the [Upload resources to remote server](../../asset/cache-manager.md) documentation.

Reviewer: Please replace all "construction" with "build".

#### Polyfills

This item is a polyfills option for some new features supported by the scripting system, and currently only supports **asynchronous functions**. After checking, the generated project will bring the corresponding polyfills, that is, the package body will be enlarged, and the developer can choose whether to use it according to actual needs.

#### Generate Immediately After Build

> If this option is checked, the **Generate** step will be executed automatically after the build is completed, and no manual operation is required.

Reviewer: ```suggestion
If this option is checked, the **Make** step will be executed automatically after the build is completed, and no manual operation is required.
```

> If this option is checked, the **Generate** step will be executed automatically after the build is completed, and no manual operation is required.

Reviewer 2: > @xunyi0 这个“生成”在编辑器里英文是“Make”，你觉得要不要改成“Generate”？

这个当时是参考了其他软件（VSCode、Xcode 等）的写法

> If this option is checked, the **Generate** step will be executed automatically after the build is completed, and no manual operation is required.

Reviewer: > > Not finished:
> > en/editor/publish/native-options.md
> > en/editor/publish/publish-web.md
> 
> What does this mean?

That means I haven't finished reviewing these two files.

> #### Encrypted script

Reviewer 2: > @xunyi0 还有这个“加密脚本”，编辑器里英文叫“Encrypt JS”，你觉得对吗？下面还有几个原文都是按照编辑器的英文来的，Jasmine给改了，你觉得要不要改呢？

按照编辑器的来吧

This item is used to encrypt the published script. The jsc file will be generated in the built-in `assets` directory, this file is encrypted. The js file will be backed up in the `script-backup` directory for debugging, and will not enter the APP during packaging.

**Script encryption key**: This value will be used as the key to encrypt the js file on the Native platform. It will be randomly generated when the project is created.

**Zip compression**: If checked, the script size can be reduced.

![encrypt js](publish-native/encrypt-js.png)

> ### Android 平台构建选项

Reviewer: Or should it be "Android Platform Build Options"? Which one is better?

> ### Android 平台构建选项

Reviewer: If you've decided which one is better, please correct all the titles similar to this one.

> Android 平台的构建选项如下：

Author: I'll change to: "Android platform build options:"

![Android build options](publish-native/android-options.png)

> #### Rendering Backend

Reviewer: ```suggestion
#### Render Backend
```

#### Game Package Name

**Game Package Name** is usually arranged in reverse order of the product website URL, such as `com.mycompany.myproduct`.

> **Note**: the package name can only contain numbers, letters and underscores. The last part of the package name must start with a letter, not an underscore or number.

#### Target API Level

Set the Target API Level required when compiling the Android platform. Click the **Set Android SDK** button next to it to quickly jump to the configuration page. For specific configuration rules, please refer to the [Configuring Native Development Environment](setup-native-development.md) documentation.

#### APP ABI

Set the CPU type that Android needs to support. It is possible to choose one or more options. Currently, there are three types: **armeabi-v7a**, **arm64-v8a**, and **x86**.

> **Note**:
>

> > 1. After you select an ABI to build, build another ABI without cleaning. At this time, the so of both ABIs will be packaged into the apk. This is the default behavior of Android Studio. If you use Android Studio to import the project, after selecting an ABI to build, first execute **Build -> Clean Project** and then build another ABI. At this time, only the latter ABI will be packaged into the apk.

Reviewer: Please correct the use of "you".

>

> > 2. After the project is imported with Android Studio, it is an independent existence and does not depend on the build release panel. If you need to modify the ABI, just modify the **PROP_APP_ABI** property in the **gradle.properties** file.

Reviewer: ```suggestion
> 2. After the project is imported with Android Studio, it is an independent existence and does not depend on the **Build** panel. If you need to modify the ABI, just modify the **PROP_APP_ABI** property in the **gradle.properties** file.
```

>
> ![modify abi](publish-native/modify_abi.png)

> #### Use the debug keystore

Reviewer: Except for the case usage, when should we keep "the" and when should we delete it in a title?

Author: you keep it, almost always. I read out loud to test it.

Reviewer: OK.

Android requires that all APKs must be digitally signed with a certificate before they can be installed. Cocos Creator provides a default keystore. Check **Use debug keystore** to use the default keystore. If the developer needs to customize the keystore, uncheck the **Use debugging keystore** check. For details, please refer to the [Official Documentation](https://developer.android.google.cn/studio/publish/app-signing).

#### Screen orientation

The screen orientation currently includes **Portrait**, **Landscape Left**, and **Landscape Right**.

> - **Portrait**: The screen is upright and the Home button is down.

Reviewer: I remember you said that the first letter after the colon should be lowercase, right?

Author: yes. fixed.

![Windows build options](publish-native/windows-options.png)

### iOS Platform Build Options

The build options for the iOS platform include **Bundle Identifier**, **Screen Orientation**, and **Rendering Backend**, where the setting of **Screen Orientation** is consistent with the Android platform.

![iOS build options](publish-native/ios-options.png)

#### Bundle Identifier

Package name, usually in reverse order of product website URL, such as `com.mycompany.myproduct`.

> **Note**: the package name can only contain numbers, letters and underscores. The last part of the package name must start with a letter, not an underscore or number.

#### Rendering Backend

**Rendering backend** currently supports **METAL**.

### Mac Platform Build Options

The build options for the Mac platform include **Bundle Identifier** and **Rendering backend**, and the setting method is the same as that of the iOS platform.

![Mac build options](publish-native/mac-options.png)

> ## Construct

Reviewer: ```suggestion
## Build
```

> After the build options are set, start building. Click the **Build** button at the bottom right corner of the **Build Release** panel to start the build process.

Reviewer: ```suggestion
After the build options are set, start building. Click the **Build** button at the bottom right corner of the **Build** panel to start the build process.
```

> When compiling scripts and packaging resources, a blue progress bar will be displayed on the **Build Tasks** page of the **Build Release** panel. If the build is successful, the progress bar will reach 100% and be displayed in green:

Reviewer: ```suggestion
When compiling scripts and packaging resources, a blue progress bar will be displayed on the **Build Tasks** page of the **Build** panel. If the build is successful, the progress bar will reach 100% and be displayed in green:
```

- If the integrated SDK is common to all native platforms, you need to add a reference in `native\engine\common\CMakeLists.txt` in the project directory.
- If you are doing secondary development based on iOS, Mac, and Windows platforms, you need to add a reference in `native\engine\currently built platform name\CMakeLists.txt` in the project directory (for example, `native\engine\ios\CMakeLists.txt) `)
- If the secondary development is based on the Android platform:
    - C++: Need to add a reference in `native\engine\android\CMakeLists.txt` in the project directory.
    - Java: Need to add a reference in the project release package `build\android\proj\build.gradle`.

For more information about the use of CMake, please refer to the [CMake Introduction](../../advanced-topics/cmake-learning.md) docuementation.

## Build and Run

Cocos Creator supports **Make** and **Run Preview** steps via the editor or the corresponding IDE for each platform (e.g.: Xcode, Android Studio, Visual Studio).

### Via the Editor

> After the build is completed, continue to click the **Generate** button next to it, and it will prompt:

Reviewer: Please replace all **Generate** button with **Make** button, and double check.

`make package YourProjectBuildPath success!`

> **Note**: after the first generation of the Android platform or the version upgrade, it is recommended to open the project through Android Studio, download the missing tools according to the prompts, and then compile and run.

**Generate** After the completion of the process, continue to click the **Run** button next to it. Part of the compilation may continue. Please wait patiently or check the progress through the log file. The operating results of each platform are:

- Mac/Windows platform will run the preview directly on the desktop
- The Android platform must be connected to the real device via USB, and the preview can be run after the USB debugging is turned on on the real device
- The iOS platform will call the simulator to run the preview, but it is recommended to connect to the real machine through Xcode to execute **Generate** and **Run**, please refer to the introduction below.

### Via IDE

> Click the **folder icon** button in the lower left corner of the **build task**, and the build release path will be opened in the file manager of the operating system. In this path, the `proj` under the `build` directory contains The currently built native platform project.

Reviewer: ```suggestion
Click the **folder icon** button in the lower left corner of the **build task**, and the build path will be opened in the file manager of the operating system. In this path, the `proj` under the `build` directory contains The currently built native platform project.
```

  ![windows xcode](publish-native/windows-vs.png)

- **iOS** and **Mac**

  ![ios xcode](publish-native/ios-xcode.png)

> **Note**: do not perform secondary development in these native platform projects, otherwise it will be overwritten when rebuilding.

> To learn how to debug on the native platform, please refer to the [Native Platform JavaScript Debugging](debug-jsb.md) documentation.

Reviewer: ```suggestion
To learn how to debug on the native platform, please refer to the [Debugging JavaScript on Native Platforms](debug-jsb.md) documentation.
```

## Precautions

1. The project built in debug mode on MIUI 10 system may pop up the prompt box of "Detected problems with API compatibility". This is a problem introduced by MIUI 10 system itself, just use release mode to build.

> 2. When packaging the iOS platform, if the developer does not use WebView related functions in the project, please make sure to remove the WebView module in **Project -> Project Settings -> Function Cropping** to improve the success of iOS App Store machine review Rate. If the developer really needs to use WebView (or the added third-party SDK comes with WebView), and therefore the iOS App Store fails the machine review, they can still try to appeal via email.

Reviewer: ```suggestion
2. When packaging the iOS platform, if the developer does not use WebView related functions in the project, please make sure to remove the WebView module in **Project -> Project Settings -> Feature Cropping** to improve the success of iOS App Store machine review Rate. If the developer really needs to use WebView (or the added third-party SDK comes with WebView), and therefore the iOS App Store fails the machine review, they can still try to appeal via email.
```

3. There are some differences between the compiled results of the Android platform through the editor and Android Studio:

>     - After executing the **make** step through the editor, the `build` directory will be generated under the release path, and the `.apk` will be generated under the `app\build\outputs\apk` directory of the `build` directory.

Reviewer: ```suggestion
    - After executing the **Make** step through the editor, the `build` directory will be generated under the release path, and the `.apk` will be generated under the `app\build\outputs\apk` directory of the `build` directory.
```

    - After compiling with Android Studio, `.apk` will be generated in `proj\app\build\outputs\apk` directory.

4. Android and Android Instant use the same build template, and the generated projects are all in the `build\android\proj` directory. Please note for this directory:

    - If it is the code used solely by the Android platform, please put it in the `app\src` directory, and put the third-party library used separately in the `app\libs` directory (if these two directories do not exist, create them).

    - If the code and third-party libraries used by Android Instant are used separately, please put them in the `instantapp\src` and `instantapp\libs` directories respectively (if these two directories do not exist, create them).

    - For code and third-party libraries shared by Android and Android Instant, please put them in the `src` and `libs` directories respectively (if these two directories do not exist, create them).

>     When compiling for Android by clicking the **Make** button in the **Build Release** panel, the default execution of `assembleRelease/Debug` will be executed, and the default execution of `instantapp:assembleRelease/Debug` when compiling Android Instant.

Reviewer: ```suggestion
    When compiling for Android by clicking the **Make** button in the **Build** panel, the default execution of `assembleRelease/Debug` will be executed, and the default execution of `instantapp:assembleRelease/Debug` when compiling Android Instant.
```

>     When you compile Android by clicking the **Generate** button in the **Build Release** panel, the default execution of `assembleRelease/Debug` will be executed, and the default execution of `instantapp:assembleRelease/Debug` when compiling Android Instant.

Reviewer: ```suggestion
    When you compile Android by clicking the **Make** button in the **Build** panel, the default execution of `assembleRelease/Debug` will be executed, and the default execution of `instantapp:assembleRelease/Debug` when compiling Android Instant.
```
